Тетрадка предназначена для студентов образвательного проекта Booster MDS использование вне проекта запрещено.

##### Задания не несут смысловой нагрузки для реальной жизни, они сделаны лишь для того, чтоб научить использовать Pandas

#### Каждое задание оценивает в 1 балл, для успешной сдачи задания нужно набрать 12 баллов

This database contains 76 attributes, but all published experiments refer to using a subset of 14 of them. In particular, the Cleveland database is the only one that has been used by ML researchers to this date. The "goal" field refers to the presence of heart disease in the patient. It is integer valued from 0 (no presence) to 1.

https://www.kaggle.com/ronitf/heart-disease-uci

___

In [1]:
import pandas as pd
import numpy as np

In [2]:
data = pd.read_csv('heart_pandas.csv')

In [4]:
data.head(5)

age     sex   chest_pain_type  resting_blood_pressure  cholesterol  \
0   63    male  non-anginal pain                     145          233   
1   37    male   atypical angina                     130          250   
2   41  female    typical angina                     130          204   
3   56    male    typical angina                     120          236   
4   57  female    typical angina                     120          354   

     fasting_blood_sugar               rest_ecg  max_heart_rate_achieved  \
0  greater than 120mg/ml                 normal                      150   
1    lower than 120mg/ml  ST-T wave abnormality                      187   
2    lower than 120mg/ml                 normal                      172   
3    lower than 120mg/ml  ST-T wave abnormality                      178   
4    lower than 120mg/ml  ST-T wave abnormality                      163   

  exercise_induced_angina  st_depression   st_slope  num_major_vessels  \
0                      no            2.3  upsloping                  0   
1                      no            3.5  upsloping                  0   
2                      no            1.4       flat                  0   
3                      no            0.8       flat                  0   
4                     yes            0.6       flat                  0   

    thalassemia  target  
0        normal       1  
1  fixed defect       1  
2  fixed defect       1  
3  fixed defect       1  
4  fixed defect       1

#### Описание признаков

**age**: The person's age in years

**sex**: The person's sex (1 = male, 0 = female)

**cp**: The chest pain experienced (Value 1: typical angina, Value 2: atypical angina, Value 3: non-anginal pain, Value 4: asymptomatic)

**trestbps**: The person's resting blood pressure (mm Hg on admission to the hospital)

**chol**: The person's cholesterol measurement in mg/dl

**fbs**: The person's fasting blood sugar (> 120 mg/dl, 1 = true; 0 = false)

**restecg**: Resting electrocardiographic measurement (0 = normal, 1 = having ST-T wave abnormality, 2 = showing probable or definite left ventricular hypertrophy by Estes' criteria)

**thalach**: The person's maximum heart rate achieved

**exang**: Exercise induced angina (1 = yes; 0 = no)

**oldpeak**: ST depression induced by exercise relative to rest ('ST' relates to positions on the ECG plot. See more here)

**slope**: the slope of the peak exercise ST segment (Value 1: upsloping, Value 2: flat, Value 3: downsloping)

**ca**: The number of major vessels (0-3)

**thal**: A blood disorder called thalassemia (3 = normal; 6 = fixed defect; 7 = reversable defect)

**target**: Heart disease (0 = no, 1 = yes)

> 1. age 
> 2. sex 
> 3. chest pain type (4 values) 
> 4. resting blood pressure 
> 5. serum cholestoral in mg/dl 
> 6. fasting blood sugar > 120 mg/dl
> 7. resting electrocardiographic results (values 0,1,2)
> 8. maximum heart rate achieved 
> 9. exercise induced angina 
> 10. oldpeak = ST depression induced by exercise relative to rest 
> 11. the slope of the peak exercise ST segment 
> 12. number of major vessels (0-3) colored by flourosopy 
> 13. thal: 3 = normal; 6 = fixed defect; 7 = reversable defect

___

**Сколько мужчин в датасете? Сколько женщен? (sex) **

In [15]:
print(len(data[data['sex'] == 'male']))
print(len(data[data['sex'] == 'female']))

207
96


In [ ]:
207
96

 **Какой процент мужчин в датасете? (решите в одну строчку, не используя предыдущий результат. Не забудте знак процента) **

In [17]:
len(data[data['sex'] == 'male'] ) / len(data) * 100

68.31683168316832

In [ ]:
68.31683168316832

**Сколько мужчин имеют заболевание сердца? Сколько женщин имеют заболевание сердца? **

In [28]:
print(len(data[(data['sex'] == 'male') & (data['target'] == 1)]))
print(len(data[(data['sex'] == 'female') & (data['target'] == 1)]))

93
72


In [18]:
93
72

**Какую долю, от общего числа пациентов, занимают мужчины не имеющие болезнь сердца?**

In [31]:
len(data[(data['sex'] == 'male') & (data['target'] == 0)])/len(data)

0.37623762376237624

In [19]:
0.37623762376237624

**Сколько лет самому молодому пациенту, страдающему болезнью сердца?**

In [41]:
data[(data['target'] == 1)].sort_values('age') 

age     sex   chest_pain_type  resting_blood_pressure  cholesterol  \
72    29    male    typical angina                     130          204   
58    34    male  non-anginal pain                     118          182   
125   34  female    typical angina                     118          210   
157   35    male    typical angina                     122          192   
65    35  female    typical angina                     138          183   
115   37  female   atypical angina                     120          215   
1     37    male   atypical angina                     130          250   
163   38    male   atypical angina                     138          175   
164   38    male   atypical angina                     138          175   
44    39    male   atypical angina                     140          321   
154   39  female   atypical angina                     138          220   
124   39  female   atypical angina                      94          199   
24    40    male  non-anginal pain                     140          199   
63    41    male    typical angina                     135          203   
133   41    male    typical angina                     110          235   
30    41  female    typical angina                     105          198   
134   41  female    typical angina                     126          306   
80    41    male   atypical angina                     112          250   
2     41  female    typical angina                     130          204   
116   41    male   atypical angina                     130          214   
122   41  female   atypical angina                     112          268   
162   41    male    typical angina                     120          157   
100   42    male  non-anginal pain                     148          244   
132   42    male    typical angina                     120          295   
22    42    male    typical angina                     140          226   
142   42  female   atypical angina                     120          209   
84    42  female    typical angina                     102          265   
149   42    male   atypical angina                     130          180   
103   42    male   atypical angina                     120          240   
74    43  female   atypical angina                     122          213   
..   ...     ...               ...                     ...          ...   
54    63  female   atypical angina                     135          252   
0     63    male  non-anginal pain                     145          233   
102   63  female    typical angina                     140          195   
13    64    male  non-anginal pain                     110          211   
139   64    male    typical angina                     128          263   
120   64  female    typical angina                     130          303   
110   64  female    typical angina                     180          325   
112   64  female   atypical angina                     140          313   
152   64    male  non-anginal pain                     170          227   
28    65  female   atypical angina                     140          417   
31    65    male    typical angina                     120          177   
38    65  female   atypical angina                     155          269   
39    65  female   atypical angina                     160          360   
150   66    male    typical angina                     160          228   
17    66  female  non-anginal pain                     150          226   
153   66  female   atypical angina                     146          278   
51    66    male    typical angina                     120          302   
127   67  female   atypical angina                     152          277   
143   67  female    typical angina                     106          223   
85    67  female   atypical angina                     115          564   
86    68    male   atypical angina                     118          277   
105   68  female   atypical an

In [ ]:
29

**Сколько лет самому возастному пациенту, у которого нет проблем с сердцем?**

In [53]:
data[(data['target'] == 0)].sort_values('age', ascending = False)

age     sex   chest_pain_type  resting_blood_pressure  cholesterol  \
238   77    male    typical angina                     125          304   
234   70    male    typical angina                     130          322   
240   70    male   atypical angina                     160          269   
225   70    male    typical angina                     145          174   
249   69    male   atypical angina                     140          254   
300   68    male    typical angina                     144          193   
203   68    male   atypical angina                     180          274   
197   67    male    typical angina                     125          254   
293   67    male   atypical angina                     152          212   
272   67    male    typical angina                     120          237   
253   67    male    typical angina                     100          299   
166   67    male    typical angina                     120          229   
165   67    male    typical angina                     160          286   
260   66  female    typical angina                     178          228   
247   66    male    typical angina                     160          246   
265   66    male    typical angina                     112          212   
199   65    male    typical angina                     110          248   
218   65    male    typical angina                     135          254   
222   65    male  non-anginal pain                     138          282   
181   65  female    typical angina                     150          225   
242   64    male    typical angina                     145          212   
233   64    male    typical angina                     120          246   
229   64    male   atypical angina                     125          309   
177   64    male   atypical angina                     140          335   
217   63    male    typical angina                     130          330   
220   63  female    typical angina                     150          407   
263   63  female    typical angina                     108          269   
296   63  female    typical angina                     124          197   
295   63    male    typical angina                     140          187   
168   63    male    typical angina                     130          254   
..   ...     ...               ...                     ...          ...   
250   51    male    typical angina                     140          298   
184   50    male    typical angina                     150          243   
188   50    male   atypical angina                     140          233   
257   50    male    typical angina                     144          200   
208   49    male   atypical angina                     120          188   
267   49    male   atypical angina                     118          149   
219   48    male    typical angina                     130          256   
171   48    male    typical angina                     110          229   
245   48    male    typical angina                     124          274   
230   47    male   atypical angina                     108          243   
274   47    male    typical angina                     110          275   
285   46    male    typical angina                     140          311   
270   46    male    typical angina                     120          249   
196   46    male   atypical angina                     150          231   
255   45    male    typical angina                     142          309   
299   45    male  non-anginal pain                     110          264   
185   44    male    typical angina                     112          290   
294   44    male    typical angina                     120          169   
200   44    male    typical angina                     110          197   
178   43    male    typical angina                     120          177   
251   43    male    typical angina                     132          247   
215   43  female    typical an

In [ ]:
77

**Сколько лет самой молодой женщине, которая страдает болезнью сердца?**

In [56]:
data[(data['sex'] == 'female') & (data['target'] == 1)].sort_values('age')

age     sex   chest_pain_type  resting_blood_pressure  cholesterol  \
125   34  female    typical angina                     118          210   
65    35  female    typical angina                     138          183   
115   37  female   atypical angina                     120          215   
154   39  female   atypical angina                     138          220   
124   39  female   atypical angina                      94          199   
2     41  female    typical angina                     130          204   
134   41  female    typical angina                     126          306   
122   41  female   atypical angina                     112          268   
30    41  female    typical angina                     105          198   
84    42  female    typical angina                     102          265   
142   42  female   atypical angina                     120          209   
74    43  female   atypical angina                     122          213   
146   44  female   atypical angina                     118          242   
53    44  female   atypical angina                     108          141   
67    45  female    typical angina                     130          234   
107   45  female    typical angina                     138          236   
94    45  female    typical angina                     112          160   
35    46  female   atypical angina                     142          177   
119   46  female    typical angina                     138          243   
118   46  female    typical angina                     105          204   
11    48  female   atypical angina                     130          275   
131   49  female    typical angina                     134          271   
135   49  female    typical angina                     130          269   
108   50  female    typical angina                     120          244   
109   50  female    typical angina                     110          254   
15    50  female   atypical angina                     120          219   
50    51  female   atypical angina                     130          256   
40    51  female   atypical angina                     140          308   
140   51  female   atypical angina                     120          295   
128   52  female   atypical angina                     136          196   
..   ...     ...               ...                     ...          ...   
59    57  female    typical angina                     128          303   
14    58  female  non-anginal pain                     150          283   
16    58  female   atypical angina                     120          340   
155   58  female    typical angina                     130          197   
89    58  female    typical angina                     100          248   
82    60  female   atypical angina                     102          318   
147   60  female  non-anginal pain                     150          240   
136   60  female   atypical angina                     120          178   
96    62  female    typical angina                     140          394   
69    62  female    typical angina                     124          209   
54    63  female   atypical angina                     135          252   
102   63  female    typical angina                     140          195   
120   64  female    typical angina                     130          303   
112   64  female   atypical angina                     140          313   
110   64  female    typical angina                     180          325   
39    65  female   atypical angina                     160          360   
38    65  female   atypical angina                     155          269   
28    65  female   atypical angina                     140          417   
17    66  female  non-anginal pain                     150          226   
153   66  female   atypical angina                     146          278   
143   67  female    typical angina                     106          223   
127   67  female   atypical an

In [22]:
34

**Какой средний возраст женщин?**

In [62]:
data[(data['sex'] == 'female')].mean()['age']

55.677083333333336

In [24]:
55.677083333333336

**Каковы средние значения и среднеквадратичные отклонения возраста тех, кто страдают болезнью сердца?**

In [67]:
print(data[data['target'] == 1].mean()['age'])
print(data[data['target'] == 1].std()['age'])

52.4969696969697
9.550650751946778


In [25]:
52.4969696969697
9.550650751946778

** Правда ли, что люди не болеющие болезнью сердца имеют уровня холестерина меньше среднего? (chol) **

In [73]:
cholesterol_not_disease = data[data['target'] == 0].mean()['cholesterol']
cholesterol_average = data.mean()['cholesterol']
cholesterol_not_disease < cholesterol_average

False

In [27]:
Неправда

**Выведите статистику rest_ecg для все числовых признаков, его максимально и среднее значение (используйте groupby(), решите в одну строчку)**

In [86]:
data.groupby('rest_ecg').agg(['max','mean'])

age            resting_blood_pressure  \
                             max       mean                    max   
rest_ecg                                                             
ST-T wave abnormality         71  52.914474                    180   
left ventricular hypertrophy  76  61.000000                    180   
normal                        77  55.687075                    200   

                                         cholesterol              \
                                    mean         max        mean   
rest_ecg                                                           
ST-T wave abnormality         129.065789         354  237.269737   
left ventricular hypertrophy  140.500000         327  261.750000   
normal                        134.027211         564  255.142857   

                             max_heart_rate_achieved              \
                                                 max        mean   
rest_ecg                                                           
ST-T wave abnormality                            194  151.960526   
left ventricular hypertrophy                     140  125.750000   
normal                                           202  147.904762   

                             st_depression           num_major_vessels  \
                                       max      mean               max   
rest_ecg                                                                 
ST-T wave abnormality                  5.6  0.879605                 4   
left ventricular hypertrophy           4.4  2.725000                 3   
normal                                 6.2  1.159184                 4   

                                       target            
                                  mean    max      mean  
rest_ecg                                                 
ST-T wave abnormality         0.638158      1  0.631579  
left ventricular hypertrophy  1.000000      1  0.250000  
normal                        0.816327      1  0.462585

In [35]:
## Ваш ответ:

**Посчитайте у кого уровень депрессии при физический нагрузке выше (в среднем), среди мужчин страдающих болезнью сердца или среди женщин не страдающих болезнью сердца (st_depression)**

In [92]:
male_depression = data[(data['sex'] == 'male') & (data['target'] == 1)]['st_depression'].mean()
female_depression = data[(data['sex'] == 'female') & (data['target'] == 0)]['st_depression'].mean()
if male_depression>female_depression:
    print('У мужчин')
else: print('У женщин')


У женщин


In [37]:
У женщин

**Посчтитайте максимальный и минимальный уровень холестерина для каждого типа chest_pain_type, rest_ecg, thalassemia. Пишите код оптимально, можно использовать циклы**

In [95]:
data.groupby('chest_pain_type').agg([max,min])['cholesterol']

max  min
chest_pain_type           
atypical angina   564  126
non-anginal pain  298  182
typical angina    409  131

In [96]:
data.groupby('rest_ecg').agg([max,min])['cholesterol']

max  min
rest_ecg                              
ST-T wave abnormality         354  126
left ventricular hypertrophy  327  197
normal                        564  149

In [98]:
data.groupby('thalassemia').agg([max,min])['cholesterol']

max  min
thalassemia                
fixed defect       417  141
normal             318  169
reversable defect  564  126

In [ ]:
## Ваш ответ:

**Сколько значений может принимать каждый из категориальных признаков?**

In [106]:
columns = ['sex', 'chest_pain_type','fasting_blood_sugar','rest_ecg','exercise_induced_angina','st_slope','thalassemia','target']
for i in columns:
    print(i + ' : ' +str(len(data[i].unique())))

sex : 2
chest_pain_type : 3
fasting_blood_sugar : 2
rest_ecg : 3
exercise_induced_angina : 2
st_slope : 2
thalassemia : 3
target : 2


In [ ]:
sex : 2
chest_pain_type : 3
fasting_blood_sugar : 2
rest_ecg : 3
exercise_induced_angina : 2
st_slope : 2
thalassemia : 3
target : 2

**У какого категориального признака наблюдается самый сильный дизбаланс классов?**

In [253]:
columns = ['sex', 'chest_pain_type','fasting_blood_sugar','rest_ecg','exercise_induced_angina','st_slope','thalassemia','target']
#я хз, можно ли на автомате искать категориальные признаки, не придумал способа, кроме как вручную забить
d = {}
length = []
kolvo_klassov = 0
spisok_znacheniy = []
disbalance = []

for i in columns:
    length.append(len(data[i].unique())) #складываем в список количество классов из каждого признака
    for j in range(len(data[i])): # формируем количество попаданий в каждый класс по каждому признаку
        if data[i][j] in d.keys():
            d[data[i][j]] += 1
        else: d[data[i][j]] = 1
    for key in d.keys():
        spisok_znacheniy.append(d[key]) # переносим количество попаданий в отдельный список и зануляем словарь. Было 
            d = {} #сделано из за того, что есть классы с одинаковыми именами, а в словаре не могут быть ключи с одинаковыми именами
    
for i in range(len(length)):
    kolvo_klassov = length[i] #количество классов для каждого признака
    for j in range(length[i]):
        test_spisok.append(spisok_znacheniy[j])#длина этого списка будет сравниваться с количеством классов
        if len(test_spisok) == kolvo_klassov:#после того, длина списка оказывается равна количеству классов
            del spisok_znacheniy[:kolvo_klassov]#мы удаляем из списка со всеми значениями уже пройденные значения
            print('Дисбаланс ' + str(i+1) + ' класса - ' + str(max(test_spisok) - min(test_spisok)))#а в тестовом смотрим разницу между min и max
            disbalance.append(max(test_spisok) - min(test_spisok))
            break
    test_spisok = []#обнуляем тестовый список, чтобы минимумы и максимумы предыдущих классов нам не мешали

print(str(columns[disbalance.index(max(disbalance))]) + ' имеет самый большой дисбаланс - ' + str(max(disbalance))) 
        

Дисбаланс 1 класса - 111
Дисбаланс 2 класса - 170
Дисбаланс 3 класса - 213
Дисбаланс 4 класса - 148
Дисбаланс 5 класса - 105
Дисбаланс 6 класса - 19
Дисбаланс 7 класса - 146
Дисбаланс 8 класса - 27
fasting_blood_sugar имеет самый большой дисбаланс - 213


In [ ]:
fasting_blood_sugar

---

@Booster
Мамкин Data Scientist

tg @mommyscience